### DISCUSSION TITLE
Potential type mismatch

### DISCUSSION THREAD

**Discussion by dibyendumajumdar** [mjhoptics/ray-optics#75] [Q&A]
I am trying to understand if there is a type mismatch here:

https://github.com/mjhoptics/ray-optics/blob/e55a4c199ddc738235813a08c562d18cce310064/src/rayoptics/raytr/analyses.py#L71

The method called returns a tuple rather than a RayPkg which is expected I think.

**Comment by mjhoptics** [comment]
The chief_ray_pkg tuple contains a RayPkg of the chief ray and a tuple that gives the exit pupil plane distance and chief ray position and direction at the exit pupil plane. It seems this isn't defined in the doc for `raytr.trace.trace_chief_ray`, or at least in nothing more than a cursory manner. The doc in `raytr.analyses.get_chief_ray_pkg` should be repeated or referenced here.

**Comment by dibyendumajumdar** [comment]
I am trying to understand what `Field.chief_ray` is supposed to hold.

For instance:

https://github.com/mjhoptics/ray-optics/blob/e55a4c199ddc738235813a08c562d18cce310064/src/rayoptics/raytr/trace.py#L458

Doesn't seem to agree with the line previously quoted

**Reply by mjhoptics** [reply]
The problem is that the function that line is contained in isn't called from anywhere. It's part of earlier work I was doing attempting to implement a wave aberration calculation due to H H Hopkins. About half the code in `raytrace.py` isn't called by any external functions.

The modules `raytrace`, `trace`, and `sequential` have some of the older, 1st generation, attempts at aberration calculations and plotting. These should be moved to a new module to avoid confusing the innocent bystander. Functions that are used to create the `RayFan`, `SpotDiagram`, and `Wavefront` figures in `axisarrayfigure` should be superseded by functions in the `analyses` module.

Sorry you got tripped up by dead code.

**Reply by dibyendumajumdar** [reply]
> Functions that are used to create the `RayFan`, `SpotDiagram`, and `Wavefront` figures in `axisarrayfigure` should be superseded by functions in the `analyses` module.

I was reviewing some of the notebooks. Is my understanding correct that these figures are superceded by new ones used in notebooks such as `focus tilt notebook.ipynb`?

**Reply by mjhoptics** [reply]
Not yet. There isn't a way (yet) to plot multiple wavelengths on a single plot. That's the reason I can recall; there may be others.

My plan would be to update the way RayFan, etc. work internally to use the API in the analyses package. I think from a user perspective, the RayFan, etc, functions have value in that they're a single function call to get a standard analysis. Sometimes you want that, sometimes you want better control over the individual plots.

**Comment by dibyendumajumdar** [comment]
Okay thank you.

I am finding it a but difficult to debug the parts that are invoked via the front-end. I presume everything in the backend can be invoked without the front-end, but I just haven't got scripts to do that yet.

**Reply by mjhoptics** [reply]
One way to search if a function is used anywhere is to search the code in the Jupyter notebook repo. If a function/method was intended for scripting, it'll probably show up there. If it's used by the rayoptics app, then the packages `gui`, `mpl`, and `qtgui` are worth surveying. The `zemax` and `codev` packages are probably useful for looking at how models in other systems are mapped onto rayoptics.
